src/number_theory/fermat_psp.lean

```diff
@@ -0,0 +1,548 @@
+  have h₁ : a - 1 ∣ a^b - 1 := sub_one_dvd_pow_sub_one _ _ (show 1 ≤ a, by linarith),
+  rw nat.div_mul_cancel h₁,
+  rw one_mul,
+
+  -- Since `a < a ^ b`, `a - 1 ≤ a ^ b - 1`
+  have h₂ : a < a^b := has_lt.lt.trans_eq' (pow_lt_pow ha hb) (pow_one a).symm,
+  show a - 1 < a^b - 1, from (tsub_lt_tsub_iff_right ha₁).mpr h₂
+end
+
+private lemma b_id_helper (a b : ℕ) (ha : 1 < a) (hb : 2 < b) : 1 < (a^b + 1)/(a + 1) :=
+begin
+  have ha₁ : 2 ≤ a := nat.succ_le_iff.mpr ha,
+  have hb₁ : 1 ≤ b := nat.one_le_of_lt hb,
+
+  -- To show that `1 < (a^b + 1) / (a + 1)`, we only need to show that `2*a + 2 ≤ (a^b + 1)`
+  suffices h : 2 ≤ (a^b + 1) / (a + 1),
+  { exact nat.succ_le_iff.mp h },
+  suffices h : 2*(a + 1) ≤ (a ^ b + 1),
+  { have h₁ : 2*(a + 1)/(a + 1) ≤ (a ^ b + 1)/(a + 1) := nat.div_le_div_right h,
+    have h₂ : 0 < a + 1 := nat.succ_pos a,
+    rwa nat.mul_div_cancel _ h₂ at h₁ },
+  rw [mul_add, mul_one],
+
+  -- Because `2 ≤ a` and `2 < b`, `3 ≤ a^(b - 1)`
+  have h₁ : 3 ≤ a^(b - 1),
+  { have : 2 ≤ b - 1 := nat.le_pred_of_lt hb,
+    calc a^(b - 1) ≥ a^2 : (pow_le_pow_iff ha).mpr this
+               ... ≥ 2^2 : pow_le_pow_of_le_left' ha₁ 2
+               ... ≥ 3   : by norm_num },
+
+  -- Since we know that `3 ≤ a ^ (b - 1)`, if we want to show `2 * a + 1 ≤ a ^ b` then it suffices
+  -- to show that `2 * a + 1 ≤ 3 * a` because then `2 * a + 1 ≤ a * 3 ≤ a * a^(b - 1) = a ^ b`
+  rw [←nat.sub_add_cancel hb₁, pow_succ],
+  suffices h : 2 * a + 1 ≤ a * a^(b - 1),
+  { exact nat.succ_le_succ h },
+  suffices h : 2 * a + 1 ≤ a * 3,
+  { exact le_mul_of_le_mul_left h h₁ },
+  rw mul_comm a 3,
+
+  -- Because `1 ≤ a`, `a + 1 ≤ 3 * a`
+  have : 3 * a = 2 * a + a := add_one_mul 2 a,
+  rw this,
+  have h : 1 ≤ a := le_of_lt ha,
+  exact add_le_add_left h (2 * a)
+end
+
+private lemma AB_id_helper (b p : ℕ) (hb : 2 ≤ b) (hp : odd p)
+  : ((b ^ p - 1) / (b - 1)) * ((b ^ p + 1) / (b + 1)) = ((b ^ (2*p)) - 1) / (b^2 - 1) :=
+have q₁ : (b - 1) ∣ (b ^ p - 1) := sub_one_dvd_pow_sub_one _ _ (show 1 ≤ b, by linarith),
+have q₂ : (b + 1) ∣ (b ^ p + 1) := begin
+  have : odd (p / 1) := eq.symm (nat.div_one p) ▸ hp,
+  have h := odd_pow_lem ↑b p 1 (one_dvd p) this,
+  rw pow_one at h,
+  exact_mod_cast h,
+end,
+have q₃ : 1 ≤ (b^p) := nat.one_le_pow p b (show 0 < b, by linarith),
+calc (b ^ p - 1) / (b - 1) * ((b ^ p + 1) / (b + 1)) =
+    (b ^ p - 1) * (b ^ p + 1) / ((b - 1) * (b + 1)) : nat.div_mul_div_comm q₁ q₂
+  ... = ((b ^ p + 1) * (b ^ p - 1)) / ((b - 1) * (b + 1)) : by rw mul_comm
+  ... = ((b ^ p)^2 - 1^2) / ((b - 1) * (b + 1))           : by rw nat.sq_sub_sq
+  ... = ((b ^ (p*2)) - 1^2) / ((b - 1) * (b + 1))         : by rw pow_mul
+  ... = ((b ^ (2*p)) - 1^2) / ((b - 1) * (b + 1))         : by rw mul_comm
+  ... = ((b ^ (2*p)) - 1^2) / ((b + 1) * (b - 1))         : by rw mul_comm (b + 1)
+  ... = ((b ^ (2*p)) - 1^2) / (b^2 - 1^2)                 : by rw nat.sq_sub_sq
+  ... = ((b ^ (2*p)) - 1) / (b^2 - 1)                     : by rw one_pow
```

Comment:
I managed to shorten your proof a little bit further by using `ring_exp` instead of `ring`, but I don't know how it performs relative to your proof.
```lean
private lemma AB_id_helper (b p : ℕ) (hb : 2 ≤ b) (hp : odd p)
  : ((b ^ p - 1) / (b - 1)) * ((b ^ p + 1) / (b + 1)) = ((b ^ (2*p)) - 1) / (b^2 - 1) :=
begin
  have q₁ : (b - 1) ∣ (b ^ p - 1) := sub_one_dvd_pow_sub_one _ _ (show 1 ≤ b, by linarith),
  have q₂ : (b + 1) ∣ (b ^ p + 1),
  { have : odd (p / 1) := eq.symm (nat.div_one p) ▸ hp,
    have h := odd_pow_lem ↑b p 1 (one_dvd p) this,
    rw pow_one at h,
    exact_mod_cast h, },
  convert nat.div_mul_div_comm q₁ q₂; rw [mul_comm (_ - 1), ←nat.sq_sub_sq],
  { ring_exp },
  { simp }
end
```